### ISSUE TITLE
Hi, Run functional test failing with an exception while unregistering the agent

### ISSUE THREAD

**Issue by Jassi1313** [Microsoft/vsts-tasks#3953]
![test](https://cloud.githubusercontent.com/assets/26910407/24730717/5e3677a0-1ab9-11e7-8bda-2d1288464055.jpg)

after finishing the test agent deployment whenever it try to run tests 
following this comes in log

2017-04-05T23:45:37.2174502Z DistributedTests: Test configuration ID: 7

10 
2017-04-05T23:45:39.0925200Z DistributedTests: Trying to unregister TestAgents from Test Machines

11 
2017-04-05T23:45:39.0955208Z DistributedTests: Task 'UnRegisterAgent' on machine 'KRKAUTOQA.branddevelopers.co.nz:5985' is being run

12 
2017-04-05T23:45:44.8007294Z DistributedTests: Task 'UnRegisterAgent' on machine 'KRKAUTOQA.branddevelopers.co.nz:5985' completed.

13 
2017-04-05T23:45:44.8017303Z DistributedTests: Trying to unregister TestAgent from Application Under Test Machines

14 
2017-04-05T23:45:44.8037311Z ##[warning]For troubleshooting, refer https://aka.ms/remotevstest.

15 
2017-04-05T23:45:44.8137386Z ##[error]System.NullReferenceException: Object reference not set to an instance of an object.

16 
2017-04-05T23:45:44.8147395Z ##[error]   at Microsoft.TeamFoundation.DistributedTask.Task.DistributedTestAutomation.DtlMachineDeploymentHelper.UnregisterAgents(String unConfigureTestAgentScriptPath)

17 
2017-04-05T23:45:44.8147395Z ##[error]   at Microsoft.TeamFoundation.DistributedTask.Task.DistributedTestAutomation.RunTests.Run()

18 
2017-04-05T23:45:44.8157403Z ##[error]   at Microsoft.TeamFoundation.DistributedTask.Task.DistributedTestAutomation.TaskRunner.Run(Dictionary`2 arguments, VssConnection connection, TaskType taskType)

19 
2017-04-05T23:45:44.8167403Z ##[error]   at Microsoft.TeamFoundation.DistributedTask.Task.DistributedTestAutomation.RundistributedTestsCmdLet.ProcessRecord()

Also the DTA Execution logs are below

I, 2008, 1, 2017/04/06, 13:44:25.610, 23341756674, DTAExecutionHost.exe, Main.TryReadAndCacheVssCredentials Acquiring the Token Info
I, 2008, 1, 2017/04/06, 13:44:25.746, 23343090374, DTAExecutionHost.exe, Communicating to the Token Servicehttps://bdltfs.visualstudio.com
W, 2008, 1, 2017/04/06, 13:44:25.832, 23343952009, DTAExecutionHost.exe, Service task failed. Retry attempt 10.
I, 8172, 1, 2017/04/06, 13:44:30.691, 23392553221, DTAExecutionHost.exe, Main.TryReadAndCacheVssCredentials Acquiring the Token Info
I, 8172, 1, 2017/04/06, 13:44:30.821, 23393830008, DTAExecutionHost.exe, Communicating to the Token Servicehttps://bdltfs.visualstudio.com
I, 8172, 1, 2017/04/06, 13:44:30.913, 23394747459, DTAExecutionHost.exe, Closing the Token Service
I, 8172, 1, 2017/04/06, 13:44:31.142, KRKAUTOQA\DTAExecutionHost, ExecutionTaskStateModel.GetExecutionAdapter : Initializing VstestAdapter.
I, 8172, 1, 2017/04/06, 13:44:31.144, KRKAUTOQA\DTAExecutionHost, ExecutionTaskStateModel.Start : Starting to monitor tasks
V, 8172, 1, 2017/04/06, 13:44:31.149, KRKAUTOQA\DTAExecutionHost, TestExecutionServiceCommandQueueListener.DispatchAsync : Getting Command for Test Agent Microsoft.TeamFoundation.Test.WebApi.TestAgent
Web method running: [https://bdltfs.visualstudio.com/_apis/test/Agents/152/Commands/0] (GET)0[test]
I, 8172, 1, 2017/04/06, 13:44:31.213, 23397753997, DTAExecutionHost.exe, ********************************************************************
I, 8172, 1, 2017/04/06, 13:44:31.213, 23397754762, DTAExecutionHost.exe, Main.InitializeAndStartExecutionHost. ExecutionHost has been started
Web method response: [https://bdltfs.visualstudio.com/_apis/test/Agents/152/Commands/0] (GET)0[test] 372 ms
V, 8172, 8, 2017/04/06, 13:44:31.780, KRKAUTOQA\DTAExecutionHost, TestExecutionServiceCommandQueueListener:DispatchAsync
W, 8172, 8, 2017/04/06, 13:44:31.784, KRKAUTOQA\DTAExecutionHost, TestExecutionServiceCommandQueueListener.DispatchAsync : TestExecutionCommand received with None. Retrying it.
W, 8172, 8, 2017/04/06, 13:44:31.784, KRKAUTOQA\DTAExecutionHost, TestExecutionServiceCommandQueueListener.DispatchAsync : Queue not found. Delaying for 3000 milliseconds.
V, 8172, 4, 2017/04/06, 13:44:34.784, KRKAUTOQA\DTAExecutionHost, TestExecutionServiceCommandQueueListener.DispatchAsync : Getting Command for Test Agent Microsoft.TeamFoundation.Test.WebApi.TestAgent
Web method running: [https://bdltfs.visualstudio.com/_apis/test/Agents/152/Commands/0] (GET)0[test]
Web method response: [https://bdltfs.visualstudio.com/_apis/test/Agents/152/Commands/0] (GET)0[test] 37 ms
V, 8172, 10, 2017/04/06, 13:44:34.822, KRKAUTOQA\DTAExecutionHost, TestExecutionServiceCommandQueueListener:DispatchAsync
W, 8172, 10, 2017/04/06, 13:44:34.822, KRKAUTOQA\DTAExecutionHost, TestExecutionServiceCommandQueueListener.DispatchAsync : TestExecutionCommand received with None. Retrying it.
W, 8172, 10, 2017/04/06, 13:44:34.823, KRKAUTOQA\DTAExecutionHost, TestExecutionServiceCommandQueueListener.DispatchAsync : Queue not found. Delaying for 9000 milliseconds.
V, 8172, 11, 2017/04/06, 13:44:43.823, KRKAUTOQA\DTAExecutionHost, TestExecutionServiceCommandQueueListener.DispatchAsync : Getting Command for Test Agent Microsoft.TeamFoundation.Test.WebApi.TestAgent
Web method running: [https://bdltfs.visualstudio.com/_apis/test/Agents/152/Commands/0] (GET)0[test]
Web method response: [https://bdltfs.visualstudio.com/_apis/test/Agents/152/Commands/0] (GET)0[test] 35 ms
V, 8172, 8, 2017/04/06, 13:44:43.859, KRKAUTOQA\DTAExecutionHost, TestExecutionServiceCommandQueueListener:DispatchAsync
W, 8172, 8, 2017/04/06, 13:44:43.859, KRKAUTOQA\DTAExecutionHost, TestExecutionServiceCommandQueueListener.DispatchAsync : TestExecutionCommand received with None. Retrying it.
W, 8172, 8, 2017/04/06, 13:44:43.859, KRKAUTOQA\DTAExecutionHost, TestExecutionServiceCommandQueueListener.DispatchAsync : Queue not found. Delaying for 27000 milliseconds.

**Comment by tanvi-soni** [comment]
Can you set system.debug=true and send us the build logs and DtaLogs at devops_tools@microsoft.com?

**Comment by Jassi1313** [comment]
Hi Tanvi,

where you want me to set system.debug=true??? In any of the config file for test agent?

Thanks
Jashwinder

**Comment by tanvi-soni** [comment]
In your build definition, under variables tab, you can add a build variable system.debug and set the value as as true. 
![image](https://cloud.githubusercontent.com/assets/13941490/24791574/46d3f2c4-1b98-11e7-9b44-c00d1715694a.png)

Queue a build with system.debug set as true and send us the build logs at devops_tools@microsoft.com

Please also send the DtaLogs and provide version of tfs, test agent.

**Comment by Jassi1313** [comment]
Hi Tanvi,
Thanks for your help. Found out that I have not assigned the associated automation to the  test case that why it was failing like that. 

I have another question for one of my release the test failed and the reason of failed test it never shows. In test result that came under others and don't know the reason why its failed.
![test](https://cloud.githubusercontent.com/assets/26910407/24840964/92d305bc-1dcc-11e7-9c2c-58a67c95ece4.jpg)

I am sending the DTA logs for this scenario to the email address that you have provided.
We are using VSTS and test agent version is 2015 with update 5 
Thanks
Regards

**Comment by tanvi-soni** [comment]
@Jassi1313  We have received your email at devops_tools@microsoft.com
I understand that your tests are failing but in test results, they are being published as "Others". Can you please verify what is the outcome of test results in trx and send that on the same thread as well.

**Comment by Jassi1313** [comment]
@tanvi-soni  trx file has not been generated for the test results. There is no trx file has been generated. Anything I can do at my end to generate the trx file 
![test](https://cloud.githubusercontent.com/assets/26910407/24879576/77744014-1e8b-11e7-94df-0d3121e41977.jpg)

**Comment by tanvi-soni** [comment]
@Jassi1313 In the attached screen shot, you have an option to filter by outcome. It is next to "group by" option which is highlighted. Filter the results by "Others". You will see all the test cases under test run link. That link will take you to the test run page from where you can download the trx. 
![image](https://cloud.githubusercontent.com/assets/13941490/24905333/3bc882aa-1ed1-11e7-9a5c-6ad6648200ab.png)

**Comment by tanvi-soni** [comment]
@Jassi1313 Can you try selecting TestAssembly option instead of test plan option and running the release. Before you run the release, go to Variables tab of release definition and set variable system.debug as true. Please send the build logs.

**Comment by Jassi1313** [comment]
@tanvi-soni  Thanks it was because two or more test cases that are associated with the same test method.
I just checked and fount out. Now its working.
Before you close this thread could you please tell me whats the best way to test run individually. I mean suppose for the overnight run as a part of release out of 15 test 2 failed and now next morning I am keen to run those 2 test again.. what would be the nice option?
2. I noticed there is no trx file generated in the test result. Is there any way to generate trx file that would be helpful . please

**Comment by tanvi-soni** [comment]
@Jassi1313  For the trx file, you can go to the test run page under Runs tab of VSTS. You will find the link to the test run in the build logs. I have highlighted the part of the log where link to test run is mentioned: 
![image](https://cloud.githubusercontent.com/assets/13941490/24945233/0bf41052-1f7c-11e7-95bb-3f9c0a7bf93a.png)

**Comment by tanvi-soni** [comment]
@Jassi1313 For the test plan/suite scenario, you can modify build definition to select only those test suites in your build definition that contain failed test and re run the build. But currently we have no other option in vNext builds that allows you to run only failed test.

**Comment by PBoraMSFT** [comment]
@Jassi1313 - for failed tests, we have the following 2 items on our backlog:
1. Provide the ability in the task to re-run failed tests a certain # of times.
2. If you are associating test automation with test cases, you can also pick failed test cases and run them on-demand from the Test hub.

**Comment by tanvi-soni** [comment]
@Jassi1313 We want to reproduce the issue that you were facing at our end. Can you provide us with trace logs from the test machine. This will greatly help us to understand the root cause of issue. You can do the following to enable the trace logs. 
i. Log into the test agent machine
ii. you will find vstest.executionengine.exe.config in following location C:\Program Files (x86)\Microsoft Visual Studio 14.0\Common7\IDE\CommonExtensions\Microsoft\TestWindow. Please set the TpTraceLevel to 4. 
Make the same changes in all vstest.executionengine.exe.config files (there will some files named as vstest.execution.x86.exe.config, etc.)
iii. Run the build and send the logs that get generated in %temp% folder

**Comment by Jassi1313** [comment]
Hi @tanvi-soni...

  Sorry for bit rate reply I was on leave. please check the attached logs.

**Comment by cltshivash** [comment]
@Jassi1313 Thanks for the logs. We are able to reproduce the issue on our end and are fixing the bug.

**Comment by Jassi1313** [comment]
Hi Prachi ,

Hope you are doing well.Just want to know whats happening with those 2
items??? are they still in your backlog or they have been implemented?

Regrads
Jashwinder

**Comment by PBoraMSFT** [comment]
@Jassi1313:
1. Yes, the re-run failed tests is on the backlog. We will be picking it up in the upcoming 3-sprint block. 

2. The ability to run tests on-demand from the Test Hub has shipped in VSTS already. You can try it out. https://www.visualstudio.com/en-us/docs/test/continuous-testing/run-automated-tests-from-test-hub

**Comment by Jassi1313** [comment]
Hi Prachi,



Have you guys released the run tests on demand. I tried the 2nd step you
mentioned but I want to run my tests on Virtual machines rather than on
local machine .Could you please help me with that.



Regards

Jashwinder

**Comment by PBoraMSFT** [comment]
@Jassi1313 
Not sure what you mean by local machine versus VMs - the steps are the same. Feel free to reach out to me directly at pbora AT microsoft DOT com and we can help you with any specific questions you may have.

**Comment by Jassi1313** [comment]
Hi Prachi..
I am not able to send the email to that address..

**Comment by Jassi1313** [comment]
Hi Prachi,
I am not able to send you the email. COuld you please help up in setting up
the running of failed tests using Team services please

**Comment by PBoraMSFT** [comment]
@Jassi1313 - not sure why you are unable to send me mail - I do receive emails from customers regularly. Are you okay with sharing your email here so I can setup the time with you?

**Comment by Jassi1313** [comment]
Hi Prachi,



Hope you are doing well. If you recall I created a task to create failed
test using Team services. SO now we have on Demand Automated Tests run that
can run failed Tests.

I Created the release definition but tests are not triggering, we are
having Coded UI Tests…

[image: cid:image001.png@01D33089.3076C820]





In original release definition for tests its trigger on the VM. When I
trigger this release will it run on the remote VM or how does it works?



In the logs of TS Run I am gettinmg following message





2017-08-27T21:47:25.7396537Z ##[section]Starting: Test run for Test plans

2017-08-27T21:47:25.7708181Z
==============================================================================

2017-08-27T21:47:25.7708181Z Task         : Visual Studio Test

2017-08-27T21:47:25.7708181Z Description  : Run tests with Visual Studio
test runner

2017-08-27T21:47:25.7708181Z Version      : 2.0.70

2017-08-27T21:47:25.7708181Z Author       : Microsoft Corporation

2017-08-27T21:47:25.7708181Z Help         : [More Information](
https://go.microsoft.com/fwlink/?LinkId=835764)

2017-08-27T21:47:25.7708181Z
==============================================================================

2017-08-27T21:47:29.8794657Z In distributed testing flow

2017-08-27T21:47:29.8794657Z
======================================================

2017-08-27T21:47:29.8794657Z Test selector : Test run

2017-08-27T21:47:29.8794657Z Test run Id : '4358'

2017-08-27T21:47:29.8794657Z Search folder :
C:\agent\_work\r42\a\NSSLobbyOarsAutomatedTest\drop\NewScriptingSystemOARS\bin\Release\NewScriptingSystemOARS.dll

2017-08-27T21:47:29.8794657Z Run settings file :
C:\agent\_work\r42\a\NSSLobbyOarsAutomatedTest\drop\DigitalTestProject\bin\Release\Digital.testsettings

2017-08-27T21:47:29.8794657Z Run in parallel : false

2017-08-27T21:47:29.8794657Z Run in isolation : false

2017-08-27T21:47:29.8794657Z Path to custom adapters : null

2017-08-27T21:47:29.8794657Z Other console options : null

2017-08-27T21:47:29.8794657Z Code coverage enabled : false

2017-08-27T21:47:29.8794657Z VisualStudio version selected for test
execution : 14.0

2017-08-27T21:47:29.9107158Z ##[error]Error: Unable to locate executable
file: 'wmic'. Please verify either the file path exists or the file can be
found within a directory specified by the PATH environment variable. Also
verify the file has a valid extension for an executable file.

2017-08-27T21:47:29.9263410Z ##[section]Finishing: Test run for Test plans





Could you please help me in regards to this  or is there any other way to
run the tests.

**Comment by tanvi-soni** [comment]
@Jassi1313 Please file a new issue for the rerun scenario since the above thread was closed post the resolution of original issue. Can you also send the release logs to devops_tools@microsoft.com

**Comment by acesiddhu** [comment]
@Jassi1313 can you create a new issue for this? by the way the issue here is that in ur releae/build definition u have a variable named PATH which is set to some value. please rename or remove that to make it work

**Comment by Jassi1313** [comment]
@Prachi Hi Prachi,
We are using Team services to run our UI functional tests on Remote
machine... We are using Deloy Test agent and than run functional tests.. As
these two task are deprecated. Is there any way to run our tests we want to
run it on our VM.

Regards
Jashwinder

On Wed, Nov 29, 2017 at 4:20 AM, Siddhartha Pandey <notifications@github.com

**Comment by PBoraMSFT** [comment]
@Jassi1313 - we are recommending that you use the VSTest task for all your testing needs. This is called out in the task UX as well as in the release notes https://review.docs.microsoft.com/en-us/vsts/release-notes/2018/jan-24-vsts?branch=master#test

**Comment by Jassi1313** [comment]
@Prachi Could you please help me to set up this for running the CodeduI
tests on virtual machines.and I cant access the link that you have sent.

Regards
Jashwinder

**Comment by Jassi1313** [comment]
@Prachi.. Hi Prachi it would be great if you could help me to set this out.
Please
